contracts/libraries/ERC20Splitter.sol

```diff
@@ -0,0 +1,131 @@
+  event Withdraw(address indexed recipient, address tokenAddress, uint256 amount);
+
+  /// @notice Deposits ERC20 or native tokens and splits between recipients based on shares.
+  /// @param tokenAddresses Array of token addresses (use address(0) for native tokens).
+  /// @param amounts Array of amounts for each token.
+  /// @param shares Array of share percentages (out of 10000) for each recipient.
+  /// @param recipients Array of recipients for each token.
+  function deposit(
+    address[] calldata tokenAddresses,
+    uint256[] calldata amounts,
+    uint256[][] calldata shares,
+    address[][] calldata recipients
+  ) external payable {
+    require(tokenAddresses.length == amounts.length, "Invalid input lengths");
+    require(tokenAddresses.length == shares.length && tokenAddresses.length == recipients.length, "Mismatched input sizes");
+
+    for (uint256 i = 0; i < tokenAddresses.length; i++) {
+      _splitTokens(tokenAddresses[i], amounts[i], shares[i], recipients[i]);
+    }
+
+    emit Deposit(msg.sender, tokenAddresses, amounts, shares, recipients);
+  }
+
+  /// @notice Withdraw all tokens that the caller is entitled to.
+  /// Tokens are automatically determined based on previous deposits.
+  function withdraw() external {
+    address[] storage userTokens = _userTokens[msg.sender];
+
+    require(userTokens.length > 0, "No tokens to withdraw");
+
+    for (uint256 i = 0; i < userTokens.length; i++) {
+      address tokenAddress = userTokens[i];
+      uint256 amount = _balances[tokenAddress][msg.sender];
+
+      if (amount > 0) {
+        _balances[tokenAddress][msg.sender] = 0;
+
+        if (tokenAddress == address(0)) {
+          payable(msg.sender).transfer(amount);
+        } else {
+          require(IERC20(tokenAddress).transfer(msg.sender, amount), "Transfer failed");
+        }
+
+        emit Withdraw(msg.sender, tokenAddress, amount);
+      }
+    }
+    _clearUserTokens(msg.sender);
+  }
+
+  /// @notice Internal function to split the tokens among recipients.
+  /// @param tokenAddress The address of the token being split (use address(0) for native tokens).
+  /// @param amount The amount of tokens to be split.
+  /// @param shares Array of share percentages (out of 10000) for each recipient.
+  /// @param recipients Array of recipients for the token.
+  function _splitTokens(address tokenAddress, uint256 amount, uint256[] calldata shares, address[] calldata recipients) internal {
+    require(shares.length == recipients.length, "Shares and recipients length mismatch");
+    require(amount > 0, "Amount must be greater than zero");
+
+    uint256 totalSharePercentage = 0;
+
+    for (uint256 i = 0; i < shares.length; i++) {
+      totalSharePercentage += shares[i];
+    }
+
+    require(totalSharePercentage == 10000, "Shares must sum to 100%");
+
+    if (tokenAddress == address(0)) {
+      // Handle native token (ETH)
+      require(msg.value == amount, "Incorrect native token amount sent");
```

Comment:
We need to discuss this. But what if multiple `address(0)` are passed? Then the `msg.value` can be greater than `amount`

This might be difficult to support though